File: problems/108_convert-sorted-array-to-binary-search-tree/main.cpp
#include <iostream>
#include <sstream>
#include <vector>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    TreeNode *build(const vector<int> &nums, int l, int r)
    {
        if (l > r)
            return nullptr;
        int mid = l + (r - l) / 2;
        TreeNode *root = new TreeNode(nums[mid]);
        root->left = build(nums, l, mid - 1);
        root->right = build(nums, mid + 1, r);
        return root;
    }

    TreeNode *sortedArrayToBST(vector<int> &nums)
    {
        return build(nums, 0, (int)nums.size() - 1);
    }
};

string vec_to_string(const vector<int> &nums)
{
    stringstream ss;
    ss << "[";
    for (size_t i = 0; i < nums.size(); ++i)
    {
        if (i > 0)
        {
            ss << ", ";
        }
        ss << nums[i];
    }
    ss << "]";
    return ss.str();
}

string preorder_with_nulls(TreeNode *root)
{
    if (!root)
    {
        return "null";
    }
    return to_string(root->val) + "," + preorder_with_nulls(root->left) + "," + preorder_with_nulls(root->right);
}

void free_tree(TreeNode *root)
{
    if (!root)
    {
        return;
    }
    free_tree(root->left);
    free_tree(root->right);
    delete root;
}

int main()
{
    Solution sol;

    vector<vector<int>> samples = {
        {-10, -3, 0, 5, 9},
        {1, 3}};
    vector<string> expected = {
        "[0,-3,9,-10,null,5] or [0,-10,5,null,-3,null,9]",
        "[3,1] or [1,null,3]"};

    for (size_t i = 0; i < samples.size(); ++i)
    {
        vector<int> nums = samples[i];
        TreeNode *result = sol.sortedArrayToBST(nums);

        cout << "Sample " << (i + 1) << ": nums = " << vec_to_string(samples[i]) << "\n";
        cout << "Placeholder result (preorder with nulls): " << preorder_with_nulls(result) << "\n";
        cout << "Expected output: " << expected[i] << "\n\n";

        free_tree(result);
    }

    return 0;
}

File: problems/108_convert-sorted-array-to-binary-search-tree/main.rs
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Box<TreeNode>>,
    pub right: Option<Box<TreeNode>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

struct Solution;

impl Solution {
    pub fn sorted_array_to_bst(nums: Vec<i32>) -> Option<Box<TreeNode>> {
        fn build(nums: &Vec<i32>, l: usize, r: usize) -> Option<Box<TreeNode>> {
            if l > r {
                return None;
            }
            let mid = l + (r - l) / 2;
            let mut root = Box::new(TreeNode::new(nums[mid]));
            if mid > 0 {
                root.left = build(nums, l, mid - 1);
            }
            root.right = build(nums, mid + 1, r);
            Some(root)
        }

        if nums.is_empty() {
            return None;
        }
        build(&nums, 0, nums.len() - 1)
    }
}

fn preorder_with_nulls(root: &Option<Box<TreeNode>>, out: &mut Vec<String>) {
    match root {
        Some(node) => {
            out.push(node.val.to_string());
            preorder_with_nulls(&node.left, out);
            preorder_with_nulls(&node.right, out);
        }
        None => out.push("null".to_string()),
    }
}

fn main() {
    let samples = vec![
        (
            vec![-10, -3, 0, 5, 9],
            "[0,-3,9,-10,null,5] or [0,-10,5,null,-3,null,9]",
        ),
        (vec![1, 3], "[3,1] or [1,null,3]"),
    ];

    for (idx, (nums, expected)) in samples.into_iter().enumerate() {
        let result = Solution::sorted_array_to_bst(nums.clone());
        let mut traversal = Vec::new();
        preorder_with_nulls(&result, &mut traversal);

        println!("Sample {}: nums = {:?}", idx + 1, nums);
        println!(
            "Placeholder result (preorder with nulls): {}",
            traversal.join(",")
        );
        println!("Expected output: {}\\n", expected);
    }
}
